File: javascript/lib/lyrics/LyricDataLoader.js
var _ = require('lodash');
var fs = require('fs');

var LyricDataLoader = function LyricDataLoader () {};

LyricDataLoader.prototype.loadDataFile = function (filename) {
    var lines = fs.readFileSync(filename, {encoding: 'utf8'}).split('\n');

    return lines.filter(function(line) {
        return line.indexOf('#') !== 0;
    });
};

LyricDataLoader.prototype.parseWordsLine = function (wordsLine) {
    return wordsLine.substr(1).split(',').filter(function(n) {return n.length}).map(function(n) {return n.toLowerCase()})
};

LyricDataLoader.prototype.parseTrackLine = function (trackLine, words) {
    var splitLine = trackLine.split(',');
    var trackId = splitLine[0];
    var mxmId = parseInt(splitLine[1], 10);
    var wordMap = {};

    _.each(_.drop(splitLine, 2), function(wc) {
        var idAndCount = wc.split(':');
        wordMap[words[parseInt(idAndCount[0], 10) - 1]] = parseInt(idAndCount[1], 10);
    });

    return {
        trackId: trackId,
        mxmId: mxmId,
        wordMap: wordMap
    }
};

LyricDataLoader.prototype.load = function (filename) {

};

module.exports = LyricDataLoader;

File: scala/src/main/scala/bbc/dojonorth/lyrics/LyricDataLoader.scala
package bbc.dojonorth.lyrics

import scala.io.Source
import scala.language.postfixOps
/**
 * This object contains code to read in the Million Song data files
 */
class LyricDataLoader {

  def load(filename: String): LyricData = {
    def lineIterator = loadDataFile(filename)

    val wordsLine = lineIterator filter (_.startsWith("%"))
    val songLines = lineIterator filterNot (_.startsWith("%"))

    val words = parseWordsLine(wordsLine.mkString(","))
    val songs = parseLinesToSongs(songLines, words)

    LyricData(words, songs)
  }

  def loadDataFile(filename: String): Iterator[String] = {
    val allLines = Source.fromFile(filename).getLines()
    allLines.withFilter { ln => !ln.startsWith("#") }
  }

  def parseWordsLine(wordsline: String): Seq[String] = {
    // drop the %, then split by comma, then lowercase
    wordsline drop(1) split(",") map(_.toLowerCase) toSeq
  }

  def parseTrackLine(trackLine: String, words: Seq[String]): SongWords = {
    val splitLine = trackLine.split(",")
    val trackId = splitLine.head
    val mxmId = splitLine.drop(1).head.toInt

    val wordIdMap = splitLine.drop(2).map { wc =>
      // word counts are specified as index to count pairs
      // the index is 1-based, so we convert to 0-based here
      val idAndCount = wc.split(":")
      words(idAndCount(0).toInt - 1) -> idAndCount(1).toInt
    } toMap

    SongWords(trackId, mxmId, wordIdMap)
  }

  def parseLinesToSongs(lines: Iterator[String], words: Seq[String]): Iterator[SongWords] =
    lines map { parseTrackLine(_, words) }
}
